src/Limiters/limiter.jl

```diff
@@ -29,99 +29,110 @@ function quasimonotone_limiter!(
-        end
-
-        local_geometry_slab = Fields.slab(space.local_geometry, e)
+    # Traverse vertical levels
+    for v in 1:vert_n_elems
+        # On each level, traverse horizontal elements
+        for he in 1:horz_n_elems
+            ρ_e_slab = Fields.slab(ρ, v, he)
+            ρq_e_slab = Fields.slab(ρq, v, he)
+            ρ_e_data = Fields.field_values(ρ_e_slab)
+            ρq_e_data = Fields.field_values(ρq_e_slab)
+            # Compute ρ's and ρq's masses over an element he
+            ρ_e_mass = sum(ρ_e_slab)
+            ρq_e_mass = sum(ρq_e_slab)
 
-        q_data .= ρq_e_data ./ ρ_e_data
+            # This should never happen, but if it does, don't limit
+            if ρ_e_mass <= zero(FT)
+                error("Negative elemental mass!")
+                continue
+            end
 
-        # Weighted least squares problem iteration loop
-        for iter in 1:(Nq * Nq - 1)
-            mass_change = zero(FT)
-            # Iterate over quadrature points
-            for j in 1:Nq, i in 1:Nq
-                ρ_wJ_data[i, j] = local_geometry_slab[i, j].WJ .* ρ_e_data[i, j]
-                # Compute the error tolerance and project q into the
-                # upper and lower bounds
-                if q_data[i, j] > max_ρq[e]
-                    mass_change += (q_data[i, j] - max_ρq[e]) * ρ_wJ_data[i, j]
-                    q_data[i, j] = max_ρq[e]
-                elseif q_data[i, j] < min_ρq[e]
-                    mass_change -= (min_ρq[e] - q_data[i, j]) * ρ_wJ_data[i, j]
-                    q_data[i, j] = min_ρq[e]
-                end
+            # Relax constraints to ensure limiter has a solution:
+            # This is only needed if running with the SSP CFL>1 or
+            # due to roundoff errors.
+            if ρq_e_mass < min_ρq[he] * ρ_e_mass
+                min_ρq[he, v] = ρq_e_mass / ρ_e_mass
             end
 
-            # By this projection of q to the upper/lower bounds, we
-            # have introduced a mass_change. Check if we are within the chosen
-            # tolerance
-            if abs(mass_change) <= rtol * abs(ρq_e_mass)
-                break
+            if ρq_e_mass > max_ρq[he] * ρ_e_mass
+                max_ρq[he, v] = ρq_e_mass / ρ_e_mass
             end
 
-            weights_sum = zero(FT)
+            local_geometry_slab = Fields.slab(space.local_geometry, he)
+
+            q_data .= ρq_e_data ./ ρ_e_data
 
-            # If the change was positive, the removed mass is added
-            if mass_change > 0
+            # Weighted least squares problem iteration loop
+            for iter in 1:(Nq * Nq - 1)
+                mass_change = zero(FT)
                 # Iterate over quadrature points
                 for j in 1:Nq, i in 1:Nq
-                    if q_data[i, j] < max_ρq[e]
-                        weights_sum += ρ_wJ_data[i, j]
+                    ρ_wJ_data[i, j] =
+                        local_geometry_slab[i, j].WJ .* ρ_e_data[i, j]
```

Comment:
What's the shape of `ρ_e_data[i, j]`? Do we need to dot operation here?

Reply:
Oh, thanks for spotting this, too! It is absolutely not necessary, and in fact, we do not use it in the analog following lines 89 and 93 below. It was just there by accident